Remove creds-init initContainer

sbwsg opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

Fixes https://github.com/tektoncd/pipeline/issues/2524 and puts us in a better position to flip the `disable-home-env-overwrite` feature flag with less fallout for users.

Fixes https://github.com/tektoncd/pipeline/issues/1271 when the feature-flag is flipped.  Also we have documentation on how to use SSH creds without the feature flag flip [in auth.md](https://github.com/tektoncd/pipeline/blob/master/docs/auth.md#using-ssh-authentication-in-your-own-git-tasks) now too.

**TLDR; The behaviour that used to be implemented by creds-init is now performed by
the entrypointer.**

Creds-init is an initContainer that writes credential files (like .ssh) to
/tekton/home. One side-effect of relying on an initContainer to write these
files is that they receive the UID / Group ID of the initContainer's process.
This side-effect breaks any Step in the Task that relies on these credentials
but runs with a different UID. An example of where this can happen is on
OpenShift, where the UID of the user is randomized for each container in
order to limit the fallout of malicious process breaking out.

This commit removes the creds-init initContainer from all TaskRun Pods. I
haven't removed the creds-init binary from our build process in this
changeset. Doing so generates a lot of extra line noise which distracts
from the core modification being presented here.

Further fallout from this change:

- Credential volume mounts from annotated Secrets used to be mounted only once,
in the `creds-init` initContainer, at `/tekcon/creds-secrets/...`. These volume mounts
are now given to every Step container so that they can be copied out by the entrypointer.
- Credentials used to be copied from `/tekton/creds-secrets/...` to `/tekton/home`
or `/tekton/creds`, depending on whether the `disable-home-env-overwrite` feature flag
was set. Now those credentials are always copied into `/tekton/creds`.
- The `$(credentials.path)` variable is now always `/tekton/creds` regardless of whether
the `disable-home-env-overwrite` feature flag is set or not.
- When the  `disable-home-env-overwrite` feature flag is set, credentials will be placed at
the actual home directory of the current user in the Step container, instead of being placed
in `/tekton/home`. If the container or Step explicitly sets a `$HOME` then credentials will
end up in there.
- Every single Step gets its own empty `/tekton/creds` volume. This allows `/tekton/creds` to be world-writable and empty for every Step to copy in to.  The directory needs to be world-writable so it's accessible even if the user's UID wouldn't normally have permissions to create a directory under `/tekton`. This is also why there are so many lines changed in the various `pkg/pod` tests.

I've put a hold on this while I add better testing and examples that exercise the cases that this PR deals with.

~~Need to figure out a safe way to introduce a credential that can be used for testing. Possibly a `read-only` Github Deploy Secret for a private test Github repo in the tektoncd org?~~

Edit: It took me way too long but I finally figured out how to spin up a git/ssh sidecar that works as an authenticated remote repo for testing creds-init.

- [x] Example that shows credentials being used in conjunction with different `securityContext`s per Step.
- [x] Example that shows using these credentials with vanilla `git` commands, not git-init or Git PipelineResource.
- [x] Documentation that describes the behaviour of credentials in the face of different `securityContext`s.

I think there also needs to be a bit of discussion about some of the more questionable design decisions.  Maybe next API WG.

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).  

# Release Notes

```release-note
Tekton's creds-init initContainer has been removed and will no longer show up in TaskRun Pods. Credentials from annotated Secrets will continue to be mounted and placed in Step containers' home directory.
```

/kind bug
/hold

sbwsg commented:
I've put a hold on this while I add better testing and examples that exercise the cases that this PR deals with.

vdemeester commented:
I am +:100: for this change :heart_eyes_cat:

chmouel commented:
I wonder if we can use a [gitolite](https://hub.docker.com/r/jgiannuzzi/gitolite/) deployment spinned just before running the tests with dynamically generated ssh keys for testing, I mean instead of having to deal with private external github repositories?

sbwsg commented:
> I wonder if we can use a gitolite deployment spinned just before running the tests with dynamically generated ssh keys for testing, I mean instead of having to deal with private external github repositories?

I've managed to spin up an example which doesn't need the private external repo - see `examples/v1beta1/taskruns/authenticating-git-commands.yaml ` in the diff here.  I'll remove that comment from the description.

sbwsg commented:
/hold cancel

OK, I think this PR is ready for review.  It may be worth taking a look at the `example` YAML I've added to get a sense of how creds-init changes in the face of different securityContexts & feature flags.

I've also added some short docs to try and help users out navigating the quagmire. It's definitely not easy to wrap one's head around unfortunately.

vdemeester commented on pkg/reconciler/taskrun/taskrun.go:
```diff
@@ -442,7 +442,7 @@ func (c *Reconciler) updateStatusLabelsAndAnnotations(tr, original *v1beta1.Task
 			err := metrics.RunningTaskRuns(c.taskRunLister)
 			if err != nil {
-				c.Logger.Warnf("Failed to log the metrics : %v", err)
+				// c.Logger.Warnf("Failed to log the metrics : %v", err)
```
I "guess" the log should still be there ? :thinking:

sbwsg replied:
:D oops, that was me trying to reduce noise while debugging unit tests.

vdemeester commented on pkg/reconciler/taskrun/controller.go:
```diff
@@ -58,7 +58,7 @@ func NewController(namespace string, images pipeline.Images) func(context.Contex
 		metrics, err := NewRecorder()
 		if err != nil {
-			logger.Errorf("Failed to create taskrun metrics recorder %v", err)
+			// logger.Errorf("Failed to create taskrun metrics recorder %v", err)
```
Same here ?

vdemeester commented:
/test pull-tekton-pipeline-integration-tests

sbwsg commented:
> We will be able to remove creds-init image in a follow-up right ?

Yes, 💯, it just made the diff really extra noisy.  I plan to move straight on to that after this.

sbwsg commented:
/test pull-tekton-pipeline-integration-tests

sbwsg commented:
Have addressed comments and improved the coverage so it's a positive rather than negative change.

GregDritschler commented on cmd/entrypoint/main.go:
```diff
@@ -41,8 +43,20 @@ var (
+	// Copy creds-init credentials from secret volume mounts to /tekton/creds
+	builders := []credentials.Builder{dockercreds.NewBuilder(), gitcreds.NewBuilder()}
+	for _, c := range builders {
+		if err := c.Write("/tekton/creds"); err != nil {
```
Why do the credentials need to be written first to /tekton/creds and then copied to $HOME?  Is there a reason they can't be written directly to $HOME?

vdemeester replied:
@GregDritschler kinda yes. `/tektoncd/creds` is the "safe" path, aka it will always be there, it is under a reserved folder, … `HOME` might not be defined in the image. Depending on the runtime, the image, … it might point to different things, it might not have the correct rights, … 

/cc @sbwsg

sbwsg replied:
Totally right @vdemeester.  We have a variable expansion, `$(credentials.path)` that has previously expanded to either `/tekton/home` (when feature flag disabled - consistent $HOME) or `/tekton/creds` (when feature flag enabled - unknowable $HOME).  After this change it always resolves to `/tekton/creds` to ensure anyone using that variable still gets a valid value that points to the credentials.

This makes me wonder, though, whether we could mount all the credentials using a single projected volume into `/tekton/creds` without the extra copy steps.  I will investigate that in a follow-up after this PR.

sbwsg replied:
And I'm going to add a comment explaining this to this file.

bobcatfish commented:
Looks like the coverage went down, is that something we want to address?

sbwsg commented:
> Looks like the coverage went down, is that something we want to address?

@bobcatfish You might be looking at an older report?  The latest numbers look positive to me:

<img width="745" alt="Screen Shot 2020-06-23 at 10 59 07 AM" src="https://user-images.githubusercontent.com/34253460/85419895-96c13400-b540-11ea-995a-a0248e5882fc.png">

bobcatfish commented:
> @bobcatfish You might be looking at an older report? The latest numbers look positive to me:

Whoops, sorry about that @sbwsg , you're totally right!

ImJasonH commented:
> Edit: It took me way too long but I finally figured out how to spin up a git/ssh sidecar that works as an authenticated remote repo for testing creds-init.

:100: I love it.

/lgtm

vdemeester commented:
I just rebase it against master and fix the conflict on the `docs/variables.md` :upside_down_face: 
/lgtm

afrittoli commented on examples/v1beta1/taskruns/authenticating-git-commands.yaml:
```diff
@@ -0,0 +1,204 @@
+      - name: messages
+        mountPath: /messages
+      script: |
+        #!/usr/bin/env ash
```
oh, never mind, I didn't know about `ash`

afrittoli commented:
/approve

afrittoli commented:
/test  pull-tekton-pipeline-integration-tests

vdemeester commented:
/retest

vdemeester commented:
/retest

vdemeester commented:
/retest
